src/games/strategy/util/UrlStreams.java

```diff
@@ -0,0 +1,87 @@
+package games.strategy.util;
+
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URL;
+import java.net.URLConnection;
+import java.util.Optional;
+import java.util.function.Function;
+
+import com.google.common.base.Throwables;
+
+import games.strategy.debug.ClientLogger;
+
+
+/**
+ * Utility class for opening input streams from URL and URI objects.
+ */
+public final class UrlStreams {
+
+  /**
+   * Opens an input stream to a given url. Returns Optional.empty() in case there is a failure.
+   * The failure message is logged to the user.
+   *
+   * @return Optional.empty() if there was a failure opening the strema, otherwise an optional
+   *         containing an input stream to the parameter uri.
+   */
+  public static Optional<InputStream> openStream(URL url) {
+    return new UrlStreams().newStream(url);
+  }
+
+
+  /**
+   * Opens an input stream to a given uri.
+   *
+   * @throws IllegalStateException if the given uri is malformed
+   *
+   * @return Optional.empty() if there was a failure opening the strema, otherwise an optional
+   *         containing an input stream to the parameter uri.
+   */
+  public static Optional<InputStream> openStream(URI uri) {
+    try {
+      return UrlStreams.openStream(uri.toURL());
+    } catch (MalformedURLException e) {
+      throw new IllegalStateException("Bad uri specified: " + uri, e);
+    }
+  }
+
+  /** Used to obtain a connection from a given URL */
+  private final Function<URL, URLConnection> urlConnectionFactory;
+
+
+  protected UrlStreams() {
+    // By default just try open a connection, raise any exceptions encountered
+    this.urlConnectionFactory = (url) -> {
+      try {
+        return url.openConnection();
```

Comment:
This code is not tested, it requires a live connection to a URL object. Instead we provide a second constructor to mock out this functionality, production code will use the no-arg version and get the default, 'live' URL connection factory 

Since this code is mostly java.lang code, and is thin, we can live without it being tested